pim-tool-back/src/main/java/vn/elca/training/service/impl/ProjectServiceImpl.java

```diff
@@ -1,30 +1,193 @@
+    private ProjectValidator projectValidator;
+
+    @Override
+    public ProjectDto create(ProjectDto projectDto) throws StartDateAfterEndDateException,
+            ProjectNumberAlreadyExistsException, InvalidProjectMembersException {
+        // Validate the input data before creating the project
+        projectValidator.validate(projectDto);
+
+        // Convert DTO to Entity
+        Project newProject = mapper.projectDtoToProject(projectDto);
+
+        // Attach group to the project
+        newProject.setGroup(groupRepository.findById(projectDto.getGroupId()).orElseThrow());
+
+        // Attach employees to the project
+        attachEmployeesToProject(newProject, projectDto.getMembers());
+
+        try {// Save the project to the database
+            Project savedProject = projectRepository.save(newProject);
+            return mapper.projectToProjectDto(savedProject);
+        } catch (DataIntegrityViolationException e) {
+            throw new ProjectNumberAlreadyExistsException("Project number already exists");
+        }
+    }
+
+    @Override
+    public ProjectDto update(ProjectDto updateProjectDto) throws StartDateAfterEndDateException,
+            ProjectNotFoundException, InvalidProjectMembersException {
+        // Validate the input data before updating the project
+        projectValidator.validate(updateProjectDto);
+
+        // Retrieve the existing project entity from the database
+        Project existingProject = projectRepository.findByProjectNumber(updateProjectDto.getProjectNumber())
+                .orElseThrow(ProjectNotFoundException::new);
+
+        // Concurrent update detection, using optimistic locking by Hibernate
+        if (existingProject.getVersion() != updateProjectDto.getVersion()) {
+            throw new ConcurrentUpdateException("Project has been updated by another user");
+        }
+
+        // Update the project with the new data
+        existingProject.setName(updateProjectDto.getName());
+        existingProject.setCustomer(updateProjectDto.getCustomer());
+        existingProject.setStatus(Project.Status.valueOf(updateProjectDto.getStatus()));
+        existingProject.setStartDate(updateProjectDto.getStartDate());
+        existingProject.setEndDate(updateProjectDto.getEndDate());
+        existingProject.setVersion(updateProjectDto.getVersion());
+
+        // Attach group to the project
+        existingProject.setGroup(groupRepository.findById(updateProjectDto.getGroupId())
+                .orElseThrow(ProjectNotFoundException::new));
+
+        // Attach employees to the project
+        attachEmployeesToProject(existingProject, updateProjectDto.getMembers());
+
+        // TODO : Hint : To handle concurrent update, the entity's status must not be persisted.
+        return mapper.projectToProjectDto(projectRepository.save(existingProject));
+    }
+
+    @Override
+    public Page<ProjectDto> findByKeyword(String keyword, String status, int page, int limit) {
```

Comment:
[N] `status` should also be enum, then you won't need to check enum.contains later.